crypto/evp/digest.c

```diff
@@ -1028,20 +1026,23 @@ int EVP_MD_up_ref(EVP_MD *md)
 {
     int ref = 0;
 
-    CRYPTO_UP_REF(&md->refcnt, &ref, md->lock);
+    if (md->origin == EVP_ORIG_DYNAMIC)
+        CRYPTO_UP_REF(&md->refcnt, &ref, md->lock);
     return 1;
 }
 
 void EVP_MD_free(EVP_MD *md)
 {
     int i;
 
-    if (md == NULL)
+    if (md == NULL || md->origin == EVP_ORIG_GLOBAL)
         return;
 
-    CRYPTO_DOWN_REF(&md->refcnt, &i, md->lock);
-    if (i > 0)
-        return;
+    if (md->origin == EVP_ORIG_DYNAMIC) {
+        CRYPTO_DOWN_REF(&md->refcnt, &i, md->lock);
+        if (i > 0)
+            return;
+    }
     ossl_provider_free(md->prov);
     CRYPTO_THREAD_lock_free(md->lock);
     OPENSSL_free(md);
```

Comment:
This function will unfortunately free `EVP_ORIG_METH` methods as well...  it's superbly easy to do that by mistake.
The proper way would be to have a common internal function that both `EVP_MD_free()` and `EVP_MD_meth_free()` call to do the actual freeing, and that they both set their conditions for calling it.  Something like this:

``` C
void EVP_MD_free(EVP_MD *md)
{
   if (md == NULL || md->origin != EVP_ORIG_DYNAMIC)
       return;

    CRYPTO_DOWN_REF(&md->refcnt, &i, md->lock);
    if (i > 0)
        return;
    evp_md_free_int(md);
}
```

``` C
void EVP_MD_meth_free(EVP_MD *md)
{
   if (md == NULL || md->origin != EVP_ORIG_METH)
       return;

    evp_md_free_int(md);
}
```

``` C
void evp_md_free_int(EVP_MD *md)
{
    ossl_provider_free(md->prov);
    CRYPTO_THREAD_lock_free(md->lock);
    OPENSSL_free(md);
}

Reply:
Your change makes sense, merging this.